pkg/utils/parse_arf_result.go

```diff
@@ -81,14 +80,15 @@ type ResourcePath struct {
 //
 //  <warning category="general" lang="en-US"><code class="ocp-api-endpoint">/apis/config.openshift.io/v1/oauths/cluster
 //  </code></warning>
-func GetPathFromWarningXML(in *xmlquery.Node) []ResourcePath {
+func GetPathFromWarningXML(in *xmlquery.Node, valuesList map[string]string) []ResourcePath {
 	apiPaths := []ResourcePath{}
 
 	codeNodes := in.SelectElements("//html:code")
 
 	for _, codeNode := range codeNodes {
 		if strings.Contains(codeNode.SelectAttr("class"), endPointTag) {
-			path := codeNode.InnerText()
+			s, _, _ := RenderValues(XmlNodeAsMarkdownPreRender(codeNode), valuesList)
+			path := strings.Replace(s, " ", "", -1)
```

Comment:
I wonder will `strings.TrimSpace(s)` works in this case?

Reply:
`strings.TrimSpace(s)` just removes beginning and ending spaces, and therefore in this case it doesn't work as expected.
```
                DumpPath: "/api/v1/namespaces/ master-mycluster1 /configmaps/ kas-config",
```

For error check I misunderstand the behaviour of `RenderValues`; I just saw the following line

https://github.com/openshift/compliance-operator/blob/f5fc6e18f72b29f42a9f9b58228aab46f269be44/pkg/utils/xml2text.go#L105

but I realized that in the following lines it returns error. I will update my code to handle these errors. Thank you for pointing out my mistake.

https://github.com/openshift/compliance-operator/blob/f5fc6e18f72b29f42a9f9b58228aab46f269be44/pkg/utils/xml2text.go#L95

https://github.com/openshift/compliance-operator/blob/f5fc6e18f72b29f42a9f9b58228aab46f269be44/pkg/utils/xml2text.go#L101

Reply:
@Vincent056 I added an error case in the test code, and I found that this modification is required to detect an error case which specifies non-existent variable.

https://github.com/openshift/compliance-operator/pull/750/files#diff-ad98feb1f5a6f396117cf1436906a17cf571de01e7a82cd4389e906d179256a9L92-R92

`make test-unit` shows no fail for this modification, but I'm wondering how this modification affects.
Is this modification acceptable?

Reply:
> 

@tmishina Thanks for the PR; sorry about the extra space, I just realized that the extra space was added by https://github.com/Vincent056/compliance-operator/blob/88e8b77ae995f93cfa484c1d39107a62f2bda3f7/pkg/utils/xml2text.go#L68, it was introduced because there aren't any spaces around variable in the rule's description and rationale, maybe we can have it removed here, and I can use a different way to add space for that maybe add in CaC before it gets parsed in Compliance Operator.

Reply:
@tmishina as far as detecting a missing value here https://github.com/openshift/compliance-operator/pull/750/files#diff-ad98feb1f5a6f396117cf1436906a17cf571de01e7a82cd4389e906d179256a9L92-R92, If I remember correctly, I think the reason it wasn't used here is that it will stop the parsing for the reset of content altogether if there is a missing value, and we don't want to have such behavior in description and rational sections. The err here was mainly for wrongly formatted content, such as `{{.so me-var}` an extra space in the content. @JAORMX suggests that we can issue an event when there is an error. Alternative to detect missing values without using `missingkey=error` https://github.com/openshift/compliance-operator/blob/fd0ab5b03eeb70ab45101b8e65518154e11730ef/pkg/utils/parse_arf_result.go#L820-L831

Reply:
@Vincent056 sorry but let me understand your suggestion...

For the first issue, can I just revert https://github.com/Vincent056/compliance-operator/blob/88e8b77ae995f93cfa484c1d39107a62f2bda3f7/pkg/utils/xml2text.go#L68 for now? or do I need to do more in addition to the revert? We also need to sync with the change on CaC as you mentioned (`I can use a different way to add space for that maybe add in CaC before it gets parsed in Compliance Operator.`).

For the second issue, how can we detect substitution errors? In case of the example you showed, it is an error case when a parsed value is not in the result value. However, in the case of this PR, not all of the variables listed in the argument are replaced, and therefore it is difficult to detect the substitution error.

Reply:
@tmishina Sorry for the confusion, for the first issue, I have a PR to fix the issue, https://github.com/openshift/compliance-operator/pull/753.

Reply:
@Vincent056 
For #1, I saw your PR; I will wait for its merger.

For #2, I'm still unclear. sorry for that...let me clarify how variable declaration is substituted to real value using `xccdf_value(varname)` defined in `macros.jinja` of `ComplianceAsCode/conent`.

1. user writes a statement like:

   ```yaml
   warning:
     general: {{{xccdf_value("myvar")}}}
   ```

2. at a time of datastream image build, ComplianceAsCode/content converts YaML into XML.

   ```xml
   <xccdf-1.2:warning xml:lang="en-US" category="general"><xccdf-1.2:sub idref="xccdf_org.ssgproject.content_value_myvar" use="legacy"/></xccdf-1.2:warning>
   ```

3. at a time of scan, Compliance Operator replaces `<xccdf-1.2:sub>` with an actual value

   ```xml
   <xccdf-1.2:warning xml:lang="en-US" category="general">myvarvalue</xccdf-1.2:warning>
   ```



You said `For the second issue, I think we should be able to detect if there is substitution error due to wrong formate` , but the input of `scan.go` (or `parse_arf_result.go`) is an XML data, and expressions like `{{varname}}` has already been replaced with XML elements by jinja of ComplianceAsCode/content. What type of format error are you expecting?

Reply:
Hi, @tmishina Do you have time for a quick google meeting, I think you might not need to change anything in CaC after all. You can just use `{{.myvar}}` in the content.

```yaml
warnings:
    - general: |-
            {{{ openshift_cluster_setting("/api/v1/namespaces/{{.myvar}}/configmaps/config") | indent(8) }}}
```
will do the trick,

Reply:
There is one step between 2 and 3,
We convert
```yaml
<xccdf-1.2:warning xml:lang="en-US" category="general"><xccdf-1.2:sub idref="xccdf_org.ssgproject.content_value_myvar" use="legacy"/></xccdf-1.2:warning>
```
into 
```yaml
<xccdf-1.2:warning xml:lang="en-US" category="general">{{.myvar}}</xccdf-1.2:warning>
```
and then we replace {{.myvar}} to the actual value in your step 3

Reply:
@Vincent056 thank you for the conversation! I revisited the code, and I realized that we need to change both `profileparser.go` and `parse_arf_result.go` because:

- Variable substitution for `description` and `rationale` is done at the time of parsing profile (by `profileparser.go`) ([code](https://github.com/openshift/compliance-operator/blob/da001d6d2fb2ace21386fb9ed7d98462bd98d324/pkg/profileparser/profileparser.go#L643)), and its output is stored as a Profile resource
- However, API resource collection is done at the time of scan by `parse_arf_result.go`, and its source is not the Profile resource but a data stream (XML) ([code](https://github.com/openshift/compliance-operator/blob/da001d6d2fb2ace21386fb9ed7d98462bd98d324/cmd/manager/scap.go#L171)), and therefore substitution at the time of parsing profile does not affect the API resource collection operation
- Hence, we need to modify:
  - `profileparser.go` to substitute variables in Profile resources
  - `parse_arf_result.go` to substitute variables in the target of API resource collection

Is my understanding correct?

Reply:
I believe that’s the correct.

Reply:
@Vincent056 pushed the revised version which modifies `parse_arf_result.go`. Important change is the following line.

https://github.com/openshift/compliance-operator/pull/750/files#diff-711414826f224eb48348b26463adec3363b3755d57196c843c06cdd9478ba338R90

now this expression is acceptable.

```
/api/v1/namespaces/{{.openshift_kube_apiserver_config_namespace}}/configmaps/{{.openshift_kube_apiserver_config_name}}
```

I will also change `profileparser.go`.

Reply:
@tmishina I think the code looks good!